File: leet_easy_c++/leet_167/leet_167b.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        unordered_map<int, int> umap;
        int tmp, dif, len = numbers.size();
        vector<int> sol(2,0);
        for(int i=0; i<len; i++){
            tmp = *(numbers.begin()+i);
            dif = target - tmp;
            if(umap.find(dif) != umap.end()){
                tmp = umap[dif] + 1;
                dif = i + 1;
                sol[0] = min(tmp, dif);
                sol[1] = max(tmp, dif);
                return sol;
            }
            umap[tmp] = i;
            tmp = *(numbers.end()-1-i);
            dif = target - tmp;
            if(umap.find(dif) != umap.end()){
                tmp = umap[dif] + 1;
                dif = len - i;
                sol[0] = min(tmp, dif);
                sol[1] = max(tmp, dif);
                return sol;
            }
            umap[tmp] = len - i - 1;
        }
        return sol;
    }
};

File: leet_easy_python/leet_167/leet_167b.py
class Solution(object):
    def twoSum(self, numbers, target):
        sz_n = len(numbers)
        for i in range(0, sz_n):
            targ, l, r = target - numbers[i], i+1, sz_n - 1
            m = (l+r)//2 
            while l<=r:
                if numbers[m] == targ: return [i+1, m+1]
                elif numbers[m] < targ: 
                    l = m+1 
                    m = (l+r)//2
                else: 
                    r = m-1
                    m = (l+r)//2
        """
        :type numbers: List[int]
        :type target: int
        :rtype: List[int]
        """

        # python binary search sol much slower than python two-pointer sol